### PR TITLE
fix: only greylist after error

### PR DESCRIPTION
**Pull request by ralph-pichler**
Only add to the greylist after actual errors. This is necessary because the handshake can and does frequently run mutiple times concurrently for the same peer.

This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2168)

### CODE REVIEW

**Review by acud — APPROVED** [review]

**Review by anatollupacescu — APPROVED** [review]

File: pkg/transaction/sender_matcher.go [hunk 1]
```diff
@@ -71,35 +71,62 @@ func (m *Matcher) Matches(ctx context.Context, tx []byte, networkID uint64, send
...
-		TimeStamp: m.timeNow(),
-		Verified:  false,
-	})
-	if err != nil {
-		return nil, err
-	}
-
 	nTx, isPending, err := m.backend.TransactionByHash(ctx, incomingTx)
 	if err != nil {
+		err2 := m.storage.Put(peerOverlayKey(senderOverlay, incomingTx), &overlayVerification{
```
Comment (esadakar) [inline pkg/transaction/sender_matcher.go:76]: maybe we can move this to a new function and call it once in a `defer` function in the case of an error?

Comment (ralph-pichler) [reply pkg/transaction/sender_matcher.go:76]: then we cannot return the `err2` to anything and we don't have logger available here without changing more stuff, tests, etc.

File: pkg/transaction/sender_matcher.go [hunk 2]
```diff
@@ -71,35 +71,62 @@ func (m *Matcher) Matches(ctx context.Context, tx []byte, networkID uint64, send
 		return nil, ErrGreylisted
```
Comment (esadakar) [inline pkg/transaction/sender_matcher.go:71]: Do we use this `Timestamp` value anywhere? I don't see a greylisting mechanism anywhere?

File: pkg/transaction/sender_matcher.go (hunk 1 above)
Comment (esadakar) [reply pkg/transaction/sender_matcher.go:76]: we can still move this out
`err2 := m.storage.Put(peerOverlayKey(senderOverlay, incomingTx), &overlayVerification{
			TimeStamp: m.timeNow(),
			Verified:  false,
		})
		if err2 != nil {
			return nil, err2
		}`

File: pkg/transaction/sender_matcher.go (hunk 2 above)
Comment (ralph-pichler) [reply pkg/transaction/sender_matcher.go:71]: Timestamp is set to `time.Now` on error. Then this if will cause `ErrGreylisted` for 5 minutes.

**Pull request merged**